bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerBaseHandler.java

```diff
@@ -21,17 +22,14 @@ public OpenSprinklerBaseHandler(Thing thing) {
     }
 
     @Override
-    public void initialize() {
-        OpenSprinklerApi ondusService = getApi();
-        if (ondusService == null) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE,
-                    "No initialized API available from bridge.");
-            return;
-        }
-        scheduler.scheduleWithFixedDelay(this::updateChannels, 0, DEFAULT_WAIT_BEFORE_INITIAL_REFRESH,
-                TimeUnit.SECONDS);
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        super.bridgeStatusChanged(bridgeStatusInfo);
 
-        updateStatus(ThingStatus.UNKNOWN);
+        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE) {
+            scheduler.scheduleWithFixedDelay(this::updateChannels, 0, DEFAULT_WAIT_BEFORE_INITIAL_REFRESH,
```

Comment:
Maybe I overlooked that, but do you cancel the scheduler if the bridge goes offline? And do you need to get a new API object here?

Reply:
I'm actually wondering why I do this at all. It's the task of the bridge to update the channels of the connected things, so that there's just one scheduler per bridge and not per thing. Additionally, the thing does not have a config for the refresh interval. I think this can be removed.

Reply:
If the bridge updates the channels you can workaround the changing API at all. But the current implementation is the things using the API they acquired before.

Reply:
Ok, this I don't get. If I see it correctly, the thing handlers always acquire the API from the bridge which is currently attached to the thing, isn't it? So there shouldn't be a case where the thing is using an old API?

Reply:
I think the real problem could be different. How is the binding recovering if the connection is lost?

Reply:
I don't think that there's a kind of "connection is lost". If the connection could not be made, the bridge should go to OFFLINE. After it comes back "online", the API client is renewed or the same, no matter what, the things will get the current, online, one from the bridge, isn't it? There's nothing saved in the thing instances, except from their configuration? Anything else is handled by the bridge.

Reply:
Let me rephrase it. If I understand your code correctly, you establish the communication in `initialize`. After that you store the API client and the bridge cones ONLINE. When child things are initialized they get exactly that client object.

Now a communication error occurs (be it a network hiccup, rebooted remote device, ...) and the bridge handler goes OFFLINE. The bridge handler informs the child things about that by calling `bridgeStatusChanged` and the things go OFFLINE, too. So far so good.

But how does the bridge reconnects? The scheduler is calling refresh, but the bridge will never return to ONLINE (same for the attached things). Shouldn‘t there be some reconnection logic involved? If e.g. the bridge is connected via a HTTP session you‘ll probably need to open a new one. `initialize` will never be called again.

I thought that this reconnection would result in a new client object. The thing handler will never receive that object, because that would only happen if `initialize` is called again. If the bridge comes ONLINE again, the thing handler is only informed via bridgeStatusChanged.

Aside from the reconnection process: you can pass the client as parameter to the refresh-method of the thing handler, that would at least make sure the thing handler is always using the most recent object.

Reply:
Ah, ok, got it! :) First of all, I think we need to look at the API client: The client has a method "openConnection", which, from the method name point of view, would indicate that it needs to be called again when the connection failed (for subsequent API calls). However, this method does not really "open" a connection in a way that it needs to be called to actually make API requests. It rather sets a specific bit in the open sprinkler device to allow some API reuests to succeed.

That means, when an API call (let's say, for turning a station on) fails, then the API call fails as well, e.g. because of a network error. The thing and the bridge may become OFFLINE.

However, the call can be retried later using the _same_ API client. The properties saved in the client are still valid (as the bridge configuration did not change, otherwise the bridge would get re-initialized). So, there's basically no need to re-initialize some objects or stuff. The polling job would simply retry to fetch the data once the interval is reached again and the call may complete then (or result in the same error).

That said, I saw some logic missing for communicating the changed state from OFFLINE to ONLINE when the refresh job succeeded, so I'll fix that :)

Reply:
Ok. According to the docs the child thing handlers are disposed before the bridge handler is disposed, so there should not be a problem in case of a changing bridge configuration.